runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -692,6 +694,33 @@ void J9::RecognizedCallTransformer::processUnsafeAtomicCall(TR::TreeTop* treetop
       if (enableTrace)
          traceMsg(comp(), "Created isObjectNull test node n%dn, non-null object will fall through to Block_%d\n", isObjectNullNode->getGlobalIndex(), treetop->getEnclosingBlock()->getNumber());
 
+      // Test if object is array - offheap only
+   #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+      if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp()->target().is64Bit())
+         {
+         //generate array check treetop
+         TR::Node *constNode = TR::Node::lconst(unsafeCall, ~(J9_SUN_FIELD_OFFSET_MASK));
+         TR::Node *unsafeAddress = TR::Node::create(TR::aladd, 2, objectNode, TR::Node::create(TR::land, 2, offsetNode, constNode));
+         TR::Node *vftLoad = TR::Node::createWithSymRef(TR::aloadi, 1, 1,
+                                                        TR::Node::createWithSymRef(unsafeAddress,
+                                                                                   comp()->il.opCodeForDirectLoad(unsafeAddress->getDataType()),
```

Comment:
directload opcode is for static field i think ... this code looks contradictory to where the addressing is from.

Reply:
I believe this is consistent with other areas where runtime array checks are performed, such as with Unsafe.get/put(): 
```
TR::Node *objLoad = TR::Node::createWithSymRef(unsafeAddress,
                                       comp()->il.opCodeForDirectLoad(unsafeAddress->getDataType()), 0,
                                       newSymbolReferenceForAddress);
TR::Node *vftLoad = TR::Node::createWithSymRef(TR::aloadi, 1, 1, objLoad,
                                       comp()->getSymRefTab()->findOrCreateVftSymbolRef());
.
.
.
TR::Node *testIsArrayFlag = comp()->fej9()->testIsClassArrayType(vftLoad);
TR::Node *isArrayNode = TR::Node::createif(TR::ificmpne, testIsArrayFlag, TR::Node::create(TR::iconst, 0), NULL);
```
(from [genCodeForUnsafeGetPut()](https://github.com/eclipse-openj9/openj9/blob/bd36b1da7b46660e9f7c38b312150b248678ee14/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L1134-L1164))

Reply:
as pointed out above, it is ops at different semantic level to me.  this caused the slew of not-understanding to begin with.